File: android/app/src/main/java/net/bradball/teetimecaddie/android/ui/common/cards/TtcCard.kt
package net.bradball.teetimecaddie.android.ui.common.cards

import android.content.res.Configuration
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Card
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import net.bradball.teetimecaddie.android.theme.TtcColorRole
import net.bradball.teetimecaddie.android.theme.MyApplicationTheme
import net.bradball.teetimecaddie.android.ui.common.icons.TtcIcons

/**
 * A card from the Fairway Morning design system — the base filled container.
 *
 * Renders a 12dp-rounded, clipped, tonal block with **no border and no shadow**: depth comes from
 * the surface ramp, not elevation. [TtcColorRole.Neutral] (the default) is the plain
 * `surfaceContainer` card; the tinted roles carry the container tones the design's inline banners
 * are built from.
 *
 * The card is a *container*, so its content is a [ColumnScope] slot and it is full width by default.
 * [contentPadding] defaults to 16dp for the common case of a plain card; pass
 * [TtcCardDefaults.NoContentPadding] for a "sectioned" card whose children must run edge to edge
 * (a header strip, a full-bleed divider, list rows that supply their own padding):
 *
 * ```
 * TtcCard(contentPadding = TtcCardDefaults.NoContentPadding) {
 *     CardHeader("8:00 AM")
 *     HorizontalDivider()
 *     PlayerRow("Brad")
 * }
 * ```
 *
 * It is a static container by default; pass [onClick] to make the whole card tappable (a game card
 * that navigates to its details, say). There is deliberately no `enabled` parameter — the design has
 * no disabled card state, as with `TtcChip`.
 *
 * @param modifier Modifier for the card.
 * @param color The [TtcColorRole] role. Defaults to [TtcColorRole.Neutral].
 * @param contentPadding Padding applied inside the card, around [content]. Defaults to
 *   [TtcCardDefaults.ContentPadding] (16dp).
 * @param onClick When non-null, makes the whole card tappable and is invoked on tap.
 * @param content The card's content, laid out in a [Column].
 */
@Composable
fun TtcCard(
    modifier: Modifier = Modifier,
    color: TtcColorRole = TtcColorRole.Neutral,
    contentPadding: PaddingValues = TtcCardDefaults.ContentPadding,
    onClick: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = TtcCardDefaults.colors(color)
    val elevation = TtcCardDefaults.elevation()
    // A card is a block-level container, so it fills its parent's width by default (matching the
    // SwiftUI side). `modifier` is applied last so a caller can still constrain the width.
    val cardModifier = Modifier.fillMaxWidth().then(modifier)

    if (onClick != null) {
        Card(
            onClick = onClick,
            modifier = cardModifier,
            shape = TtcCardDefaults.Shape,
            colors = colors,
            elevation = elevation,
        ) {
            Column(Modifier.padding(contentPadding), content = content)
        }
    } else {
        Card(
            modifier = cardModifier,
            shape = TtcCardDefaults.Shape,
            colors = colors,
            elevation = elevation,
        ) {
            Column(Modifier.padding(contentPadding), content = content)
        }
    }
}

// ─── Preview-only stand-ins ──────────────────────────────────────────────────
// The real CardHeader / ListRow / TtcDivider components come in a later pass; these exist only so
// the sectioned preview below can prove the container composes correctly.

@Composable
private fun PreviewCardHeader(title: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceContainerHigh)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            painter = TtcIcons.CALENDAR_CLOCK.painter,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(20.dp),
        )
        Spacer(Modifier.width(10.dp))
        Text(title, style = MaterialTheme.typography.titleMedium)
    }
}

@Composable
private fun PreviewListRow(primary: String, secondary: String? = null) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            painter = TtcIcons.PERSON.painter,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(22.dp),
        )
        Spacer(Modifier.width(14.dp))
        Column {
            Text(primary, style = MaterialTheme.typography.bodyLarge)
            if (secondary != null) {
                Text(
                    secondary,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

@Preview(name = "Light", showBackground = true)
@Preview(name = "Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcCardPreview() {
    MyApplicationTheme {
        Surface {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                TtcCard {
                    Text("Filled card", style = MaterialTheme.typography.titleMedium)
                    Text(
                        "surfaceContainer, 12dp radius, no border, no shadow.",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                TtcCard(color = TtcColorRole.Primary) { Text("Primary container") }
                TtcCard(color = TtcColorRole.Secondary) { Text("Secondary container") }
                TtcCard(color = TtcColorRole.Tertiary) { Text("Tertiary container") }
                TtcCard(color = TtcColorRole.Error) { Text("Error container") }
                TtcCard(onClick = {}) { Text("Tappable card") }
            }
        }
    }
}

@Preview(name = "Sectioned - Light", showBackground = true)
@Preview(name = "Sectioned - Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcCardSectionedPreview() {
    MyApplicationTheme {
        Surface {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                // A "list card": no card padding, so the header and dividers run edge to edge.
                TtcCard(contentPadding = TtcCardDefaults.NoContentPadding) {
                    PreviewCardHeader("8:00 AM")
                    PreviewListRow("Brad", "Organizer")
                    HorizontalDivider(modifier = Modifier.padding(start = 52.dp))
                    PreviewListRow("Jeff", "Confirmed")
                }

                // The shape an inline banner will take once its content component exists.
                TtcCard(color = TtcColorRole.Error) {
                    Text("We don't recognize that email", style = MaterialTheme.typography.titleSmall)
                    Text(
                        "There's no TeeTimeCaddie account for casey@golf.app.",
                        style = MaterialTheme.typography.bodyMedium,
                    )
                }
            }
        }
    }
}

File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/ui/common/Cards/TtcCard.swift
import SwiftUI
import ThemeUI

/// A card from the Fairway Morning design system — the base filled container.
///
/// Renders a `radius-md`-rounded, clipped, tonal block with **no border and no shadow**: depth comes from
/// the surface ramp, not elevation. `.neutral` (the default) is the plain `surfaceContainer` card;
/// the tinted ``TtcColorRole`` cases carry the container tones the design's inline banners are built
/// from.
///
/// The card is a *container*, so its content is a `@ViewBuilder` slot laid out in a leading-aligned
/// `VStack` (matching Compose's `ColumnScope`, so the content block reads identically on both
/// platforms) and it is full width by default. `contentPadding` defaults to 16pt for the common case
/// of a plain card; pass ``TtcCardStyle/noContentPadding`` for a "sectioned" card whose children must
/// run edge to edge — a header strip, a full-bleed divider, or list rows that supply their own
/// padding:
///
/// ```swift
/// TtcCard(contentPadding: TtcCardStyle.noContentPadding) {
///     CardHeader("8:00 AM")
///     TtcDivider()
///     PlayerRow("Brad")
/// }
/// ```
///
/// It is a static container by default; pass `onClick` to make the whole card tappable (a game card
/// that navigates to its details, say). There is deliberately no disabled state — the design has
/// none, as with ``TtcChip``.
struct TtcCard<Content: View>: View {
    @EnvironmentObject private var theme: AppTheme

    private let color: TtcColorRole
    private let contentPadding: EdgeInsets
    private let onClick: (() -> Void)?
    private let content: Content

    init(
        color: TtcColorRole = .neutral,
        contentPadding: EdgeInsets = TtcCardStyle.contentPadding,
        onClick: (() -> Void)? = nil,
        @ViewBuilder content: () -> Content
    ) {
        self.color = color
        self.contentPadding = contentPadding
        self.onClick = onClick
        self.content = content()
    }

    /// The card chrome: the content column, inset by `contentPadding`, on the role's container fill,
    /// clipped to the card shape.
    private var surface: some View {
        let colors = TtcCardStyle.colors(color, theme.colorScheme)
        return VStack(alignment: .leading, spacing: 0) { content }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(contentPadding)
            .background(colors.background)
            .foregroundStyle(colors.foreground)
            .clipShape(TtcCardStyle.shape(theme.shapes))
    }

    var body: some View {
        if let onClick {
            Button(action: onClick) { surface }
                .buttonStyle(TtcCardButtonStyle(shape: TtcCardStyle.shape(theme.shapes)))
        } else {
            surface
        }
    }
}

#Preview("Light") {
    TeeTimeCaddieTheme {
        TtcCardPreviewContent()
    }
}

#Preview("Dark") {
    TeeTimeCaddieTheme {
        TtcCardPreviewContent()
    }
    .preferredColorScheme(.dark)
}

private struct TtcCardPreviewContent: View {
    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                TtcCard {
                    Text("Filled card").font(.headline)
                    Text("surfaceContainer, radius-md, no border, no shadow.")
                        .font(.subheadline)
                }
                TtcCard(color: .primary) { Text("Primary container") }
                TtcCard(color: .secondary) { Text("Secondary container") }
                TtcCard(color: .tertiary) { Text("Tertiary container") }
                TtcCard(color: .error) { Text("Error container") }
                TtcCard(onClick: {}) { Text("Tappable card") }

                // A "list card": no card padding, so the header and dividers run edge to edge.
                TtcCard(contentPadding: TtcCardStyle.noContentPadding) {
                    PreviewCardHeader(title: "8:00 AM")
                    PreviewListRow(primary: "Brad", secondary: "Organizer")
                    PreviewRowDivider(inset: 52)
                    PreviewListRow(primary: "Jeff", secondary: "Confirmed")
                }

                // The shape an inline banner will take once its content component exists.
                TtcCard(color: .error) {
                    Text("We don't recognize that email").font(.subheadline.weight(.semibold))
                    // `Text(verbatim:)` — a plain `Text` markdown-parses its key and would
                    // auto-link the email, overriding the card's foreground color.
                    Text(verbatim: "There's no TeeTimeCaddie account for casey@golf.app.")
                        .font(.footnote)
                }
            }
            .padding()
        }
    }
}

// ─── Preview-only stand-ins ──────────────────────────────────────────────────
// The real CardHeader / ListRow / TtcDivider components come in a later pass; these exist only so
// the sectioned preview above can prove the container composes correctly.

private struct PreviewCardHeader: View {
    @EnvironmentObject private var theme: AppTheme
    let title: String

    var body: some View {
        let scheme = theme.colorScheme
        return HStack(spacing: 10) {
            Image(systemName: "clock.fill").foregroundStyle(scheme.primary)
            Text(title).font(.headline)
            Spacer()
        }
        .padding(EdgeInsets(top: 12, leading: 16, bottom: 12, trailing: 16))
        .frame(maxWidth: .infinity)
        .background(scheme.surfaceContainerHigh)
    }
}

private struct PreviewListRow: View {
    @EnvironmentObject private var theme: AppTheme
    let primary: String
    let secondary: String?

    var body: some View {
        let scheme = theme.colorScheme
        return HStack(spacing: 14) {
            Image(systemName: "person.fill").foregroundStyle(scheme.onSurfaceVariant)
            VStack(alignment: .leading, spacing: 1) {
                Text(primary).font(.body)
                if let secondary {
                    Text(secondary).font(.caption).foregroundStyle(scheme.onSurfaceVariant)
                }
            }
            Spacer()
        }
        .padding(EdgeInsets(top: 12, leading: 16, bottom: 12, trailing: 16))
    }
}

private struct PreviewRowDivider: View {
    @EnvironmentObject private var theme: AppTheme
    let inset: CGFloat

    var body: some View {
        Rectangle()
            .fill(theme.colorScheme.outlineVariant)
            .frame(height: 1)
            .padding(.leading, inset)
    }
}
